api/websocket/server/relay.go

```diff
@@ -124,11 +125,44 @@ func (ws *WsServer) sendInboundRelayMessage(relayMessage *pb.Relay) {
 				sigChainLen: int(relayMessage.SigChainLen),
 			})
 		}
-	} else {
+		if time.Duration(relayMessage.MaxHoldingSeconds) > pongTimeout/time.Second {
+			ok := ws.messageDeliveredCache.Push(relayMessage)
+			if !ok {
+				log.Warningf("MessageDeliveredCache full, discarding messages.")
+			}
+		}
+	} else if relayMessage.MaxHoldingSeconds > 0 {
 		ws.messageBuffer.AddMessage(clientID, relayMessage)
 	}
 }
 
+func (ws *WsServer) startCheckingLostMessages() {
+	for {
+		v, ok := ws.messageDeliveredCache.Pop()
+		if !ok {
+			break
+		}
+		if relayMessage, ok := v.(*pb.Relay); ok {
+			clientID := relayMessage.DestId
+			clients := ws.SessionList.GetSessionsById(hex.EncodeToString(clientID))
+			if len(clients) > 0 {
+				threshold := time.Now().Add(-pongTimeout)
+				success := false
+				for _, client := range clients {
+					if client.GetLastReadTime().After(threshold) {
+						success = true
+						break
+					}
+				}
+				if success {
+					continue
+				}
```

Comment:
What happen if a multi-dest msg had received by part of clients?
This piece looks that the msg will be marked **success** and skip to next pop() if one of dest client `GetLastReadTime` after `threshold `.

Reply:
There is no such thing called multi-dest msg, each msg only has one dest. A multi-dest msg will be separated into multiple msgs.